zaza/utilities/run_report.py

```diff
@@ -0,0 +1,196 @@
+
+    :param cloud_name: Name of cloud eg canonistack1
+    :type cloud_name: str
+    :param model_name: Name of model
+    :type model_name: str
+    :param target_bundle: Name of bundle
+    :type target_bundle: str
+    """
+    global run_data
+    _init_run_data()
+    if cloud_name:
+        run_data[ReportKeys.METADATA]['cloud_name'] = cloud_name
+    if model_name:
+        run_data[ReportKeys.METADATA]['model_name'] = model_name
+    if target_bundle:
+        run_data[ReportKeys.METADATA]['target_bundle'] = target_bundle
+
+
+def get_events_start_stop_time(events):
+    """Return the time of the first event and the last.
+
+    :param events: Dict of events.
+    :type events: dict
+    :returns: Tuple of start time and endtime in seconds since epoch.
+    :rtype: (float, float)
+    """
+    start_time = -1
+    finish_time = -1
+    for event_name, event_info in events.items():
+        for event_state, timestamp in event_info.items():
+            if start_time == -1 or start_time > timestamp:
+                start_time = timestamp
+            if finish_time == -1 or finish_time < timestamp:
+                finish_time = timestamp
+    return start_time, finish_time
+
+
+def get_event_report():
+    """Produce report based on current run.
+
+    :returns: Dictionary of run data with summary info.
+    :rtype: dict
+    """
+    global run_data
+    _init_run_data()
+    report = copy.deepcopy(run_data)
+    start_time, finish_time = get_events_start_stop_time(get_events())
+    full_run_time = finish_time - start_time
+    for name, info in report[ReportKeys.EVENTS].items():
+        try:
+            event_time = (info[EventStates.FINISH] -
+                          info[EventStates.START])
+            info[ReportKeys.ELAPSED_TIME] = event_time
+            info[ReportKeys.PCT_OF_RUNTIME] = math.ceil(
+                (event_time / full_run_time) * 100)
+        except KeyError:
+            pass
+    return report
+
+
+def write_event_report(output_file=None):
+    """Log the event report and optionally write to a file.
+
+    :param outputfile: File name to write summary to.
+    :type events: str
+    """
+    report = get_event_report()
+    report_txt = yaml.dump(
+        report,
+        Dumper=EnumToStrDumper,
+        default_flow_style=False)
+    if output_file:
+        with open(output_file, 'w') as out:
+            out.write(report_txt)
+    logging.info(report_txt)
+
+
+def get_run_data():
+    """Return the raw dictionary of events.
+
+    :returns: Dictionary of run data without summary info.
+    :rtype: dict
+    """
+    global run_data
+    _init_run_data()
+    return run_data
+
+
+def reset_run_data():
+    """Reset the run data, removing all entries."""
+    _init_run_data(reset=True)
+
+
+def get_events():
+    """Return the dictionary of events recorded for this run.
+
+    :returns: Dictionary of events.
+    :rtype: dict
+    """
+    return get_run_data()[ReportKeys.EVENTS]
```

Comment:
Should this be a copy?  i.e. maybe the function names should be explicit when a mutable ref is returned vs a full copy of the data; how is `get_events()` being used?